### DISCUSSION TITLE
Feature Request: Helpful details for building bots for PR-related actions

### DISCUSSION THREAD

**Discussion by freakboy3742** [community/community#26109] [Actions]
<p>I’ve been porting some CI tooling to GitHub Actions; the experience has been pretty good overall, but I’ve hit three rough edges that would be nice to be smoothed out. The first two have workarounds, but they aren’t great; a baked-in solution would be much nicer.</p>
<p><strong>1. An Action that triggers on a Pull Request can’t easily get the PR number.</strong></p>
<p>If you have an action triggering on a Pull Request, you may need to hit the Github API to get details about the PR. To do this, you need to know the PR number; however, that value isn’t present in the environment.</p>
<p>You *can* extract this value by parsing ${{ github.ref }} (e.g., with $(cut -d’/’ -f3 &lt;&lt;&lt; ${{ github.ref }})) - however, this seems an overly messy way of getting something that is a fundamental property of the event that triggered the action.</p>
<p><strong>2. An Action that triggers on a Pull Request can’t easily get the hash that triggered the Action</strong></p>
<p>When a PR triggers an action, the repository that you receive as a checkout *isn’t* at the SHA hash of the commit that stimulated the PR - it’s the SHA of the *merge commit* that would be generated if the PR was accepted. This is the SHA that is reported as ${{ github.sha }} However, that SHA *isn’t* part of the repository until the PR is merged; so if you want to do something like post a comment on the PR, you have to reverse engineer a hash that *is* currently part of the repository.</p>
<p>As a workaround, you can extract the SHA of the parent of the merge commit (e.g, with $(cut -d’ ’ -f3 &lt;&lt;&lt; $(git rev-list --parents -n 1 HEAD)) - but again, this is a somewhat complex and brittle way to extract something that is a fundamental property of the pull request.<br></p>
<p><strong>3. All comments posted by an Action are posted as the github-actions user</strong></p>
<p>If you’re on an Open Source project, almost <em>none</em> of your contributions are internal; they’re <em>all</em> from forks. Some Open Source projects also make extensive use of bots to automate common actions (such as providing automated code review comments).</p>
<p>However, Actions don’t allow secrets to be exposed to pull requests originated from forks. You <em>can</em> use the GITHUB_TOKEN - but all actions performed using that token are performed by the “github-actions” user. </p>
<p>If you have multiple bot actions that you need to be performend, it can be helpful for those bots to have different identities (e.g., one bot that welcomes new contributors, a second providing code review, etc); however, being forced into a single auth token mean you lose any ability to expose unique identities.<br></p>
<p>I don’t have a workaround for this last one, other than just accepting that all actions will be performed by the github-actions user. Having the ability to nominate the user/bot whose credentials will be used when running an action would be very helpful.</p>

**Comment by BelHFL** [comment]
<p>Hi <a class="mention" href="/freakboy3742">@freakboy3742</a> ,</p>
<p>Thanks for this feedback! We’re always working to improve GitHub and the GitHub Community Forum, and we consider every suggestion we receive. I’ve logged your feature request in our internal feature request list. Though I can’t guarantee anything or share a timeline for this, I can tell you that it’s been shared with the appropriate teams for consideration.</p>
<p>Cheers!</p>

**Comment by chrispat** [comment, marked as answer]
<p><a class="mention" href="/freakboy3742">@freakboy3742</a> </p>
<p>One of the things GitHub Actions provides is the full payload for the event that triggered the workflow run and you can access that payload using the ${{ }} syntax.  As mentioned in the <a href="https://help.github.com/en/articles/contexts-and-expression-syntax-for-github-actions" rel="nofollow noopener">context document</a> you can access the event payload through the <a href="https://help.github.com/en/articles/contexts-and-expression-syntax-for-github-actions#github-context" rel="nofollow noopener">github context</a>.  You can navigate into the contents of the JSON object from the event payload by simply accessing the property.</p>
<p>The pull_request event: <a href="https://developer.github.com/v3/activity/events/types/#pullrequestevent" rel="nofollow noopener">https://developer.github.com/v3/activity/events/types/#pullrequestevent</a></p>
<p>For example, to get the pull request number you would use ${{ github.event.number }}.  To get the head sha you would use ${{ github.event.pull_request.head.sha }}</p>
<p>As for different users for different workflows I am not sure I have a great solution right now.  It is something we will have to think about.</p>

**Comment by freakboy3742** [comment]
<p><span class="mention">@chrispatThanks</span> - very helpful! It was surprising to me that this detail wasn’t available; nice to see it was an oversight on my part.</p>
<p>By way of follow up suggestion: I think my oversight stemmed from the fact that the connection from the documentation that describes the context data (which I *did* find) to the documentation for the structure of the event data wasn’t entirely clear.</p>
<p>The “event description” in the table at <a href="https://help.github.com/en/articles/events-that-trigger-workflows#webhook-events" rel="nofollow noopener">https://help.github.com/en/articles/events-that-trigger-workflows#webhook-events</a> points to <a href="https://developer.github.com/v3/pulls/" rel="nofollow noopener">https://developer.github.com/v3/pulls/</a> (which is a very high level description of interacting with pull requests) rather than <a href="https://developer.github.com/v3/activity/events/types/#pullrequestevent" rel="nofollow noopener">https://developer.github.com/v3/activity/events/types/#pullrequestevent</a> which is a direct description of the event that will be received. It appears that this is a common pattern in that table; the link (if provided) is to a description of an API endpoint that would cause the action, rather than the definition of the event that will be received.</p>
<p>Might I suggest updating the links in that table to point at the relevant event definition, rather than the API endpoint? Or adding a separate column with a link to the event data definition?</p>
